k2/csrc/context_inl.cuh

```diff
@@ -0,0 +1,182 @@
+// handle different GPU devices on multiple machines, that's also the reason
+// that we pass `Context` instead of `DeviceType` as the input parameter here.
+inline MemoryCopyKind GetMemoryCopyKind(const Context &src,
+                                        const Context &dst) {
+  if (src.GetDeviceType() == kCpu && dst.GetDeviceType() == kCpu) {
+    return MemcpyHostToHost;
+  } else if (src.GetDeviceType() == kCpu && dst.GetDeviceType() == kCuda) {
+    return MemcpyHostToDevice;
+  } else if (src.GetDeviceType() == kCuda && dst.GetDeviceType() == kCpu) {
+    return MemcpyDeviceToHost;
+  } else if (src.GetDeviceType() == kCuda && dst.GetDeviceType() == kCuda) {
+    return MemcpyDeviceToDevice;
+  } else {
+    K2_LOG(FATAL) << "Unsupported Context";
+    return MemcpyUnknown;
+  }
+}
+
+inline void MemoryCopy(void *dst, const void *src, std::size_t count,
+                       MemoryCopyKind kind) {
+
+  std::map<MemoryCopyKind, cudaMemcpyKind> copy_kind_mappings = {
+      {MemcpyHostToHost, cudaMemcpyHostToHost},
+      {MemcpyHostToDevice, cudaMemcpyHostToDevice},
+      {MemcpyDeviceToHost, cudaMemcpyDeviceToHost},
+      {MemcpyDeviceToDevice, cudaMemcpyDeviceToDevice}};
+  auto it = copy_kind_mappings.find(kind);
+  K2_CHECK_NE(it, copy_kind_mappings.end());
+  cudaMemcpy(dst, src, count, it->second);
+}
+
+template <typename T1, typename T2>
+bool IsCompatible(const T1 &t1, const T2 &t2) {
+  // suppose both T1 and T2 have member method `Context`
+  return t1.Context()->IsCompatible(*t2.Context());
+}
+
+template <typename T>
+ContextPtr GetContext(T &t) {
+  // suppose T has member method `Context`
+  ContextPtr c;
+  t.To(c);
+  return c;
+}
+
+template <typename First, typename... Rest>
+ContextPtr GetContext(First &first, Rest &... rest) {
+  ContextPtr ans1 = GetContext(first), ans2 = GetContext(rest...);
+  K2_DCHECK(ans1->IsCompatible(*ans2)) << "Contexts are not compatible";
+  return ans1;
+}
+
+/*
+  Convenience wrapper for NewRegion() that takes the context from a provided
+  region.
+ */
+inline std::shared_ptr<Region> NewRegion(Region &region, std::size_t num_bytes) {
+  return NewRegion(region.context, num_bytes);
+}
+
+// Objects from k2 generally have a Context() method, so this template
+// will work to get the device-type for pretty arbitrary objects.
+template <typename T>
+inline DeviceType DeviceOf(const T &t) {
+  return t.Context()->GetDeviceType();
+}
+
+/* Eval() will evaluate lambda(i) for 0 <= i < n, on the appropriate
+   device (CPU or GPU). */
+template <typename LambdaT>
+void Eval(cudaStream_t stream, int32_t n, LambdaT &lambda) {
+  if (n <= 0) return;  // actually it would be an error if n < 0.
+  if (stream == kCudaStreamInvalid) {
+    // TODO: if n is very large, we'll eventually support running this with
+    // multiple threads.
+    for (int32_t i = 0; i < n; ++i) {
+      lambda(i);
+    }
+  } else {
+#ifdef __CUDA_ARCH__
```

Comment:
This file has suffix `.cuh` so I think it will be processed by nvcc. What will happen if we remove this macro guard here?

Reply:
Actually this macro guard (#ifdef __CUDA_ARCH__) is not correct, I think.  This is host code, so __CUDA_ARCH__ would always be undefined when evaluating it.  (It is a kernel invocation, but that's still host code).